runtimes/rust/src/spec.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2020 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Specifications of Rust and non-Rust artifacts for use in deployment.
+
+use exonum::{
+    blockchain::config::{GenesisConfigBuilder, InstanceInitParams},
+    merkledb::BinaryValue,
+    runtime::{migrations::MigrateData, ArtifactId, InstanceId, RuntimeIdentifier},
+};
+
+use std::{marker::PhantomData, mem};
+
+use self::sealed::Sealed;
+use crate::{DefaultInstance, RustRuntimeBuilder, ServiceFactory};
+
+mod sealed {
+    pub trait Sealed {}
+}
+
+/// Denotes a data type that can be used as an argument to `NodeBuilder::with()`.
+///
+/// This is a sealed trait: it is not meant to be implemented by the external data types.
+pub trait Deploy: Sealed {
+    /// Modifies the genesis config and Rust runtime builders to deploy self.
+    fn deploy(self, genesis: &mut GenesisConfigBuilder, rt: &mut RustRuntimeBuilder);
+}
+
+/// Marker type for artifact deployment without migration support.
+#[derive(Debug)]
+pub struct Simple(());
```

Comment:
Why not just `Service`?

Reply:
I think `Simple` / `Migrating` dichotomy provides a better description than `Service` / `Migrating`. (For one, "simple" and "migrating" are both adjectives 🙃)

Reply:
Oh, that is my mistake, I meant that we can just write `struct Simple` instead of `struct Simple(())`.

Reply:
I wanted `Simple` / `Migrating` to be marker types (i.e., not instantiable).